torch/jit.py

```diff
@@ -53,224 +523,61 @@ def _fork_rng(enabled=True):
-        trace_inputs = self.get_trace_inputs(args)
-        trace_info = self.traces.get(self.get_input_key(trace_inputs))
-        if trace_info is None:
-            return False
-        trace_info.check_traces()
-        return trace_info.complete_trace is not None
-
-    def __call__(self, *args):
-        # Run the real thing if tracing is disabled
-        if not self.enabled:
-            with _time("run_real", self.time):
-                return self._run(*args)
-
-        trace_inputs = self.get_trace_inputs(args)
-        input_key = self.get_input_key(trace_inputs)
-        trace_info = self.traces[input_key]
-        # Use the compiled closure if we have it already
-        if trace_info.closure is not None:
-            return self.run_closure(trace_info, args, trace_inputs)
-
-        # Check if any of the traces in our pool are complete now
-        trace_info.check_traces()
-        if trace_info.complete_trace is not None:
-            trace_info.compile_trace(self.optimize)
-            return self.run_closure(trace_info, args, trace_inputs)
-
-        # Otherwise, we have to collect a new trace
-        trace, out = self.record_trace(args)
-        trace_info.traces.append(trace)
-        if trace_info.proto is None:
-            trace_info.proto = function._to_proto(out)
-        return out
-
-
-def record_trace(traceable, *args, **kwargs):
-    """
-    Record a trace for a traceable object (either a function or a Module),
-    returning a tuple (trace, output).  Positional arguments are passed
-    as arguments to the model, while keyword arguments are used to control
-    how we go about performing the trace.
-
-    TODO: document kwargs
-    """
-    parameters = kwargs.pop('parameters', ())
-    return Traceable(traceable, **kwargs).record_trace(args, extra=parameters)
-
-
-def traced(traceable, **traced_kwargs):
-    t = Traceable(traceable, **traced_kwargs)
-    if isinstance(traceable, Module):
-        traceable.forward = t
-        return traceable
-    else:
-        return t
+    filename = "{}_{}".format(trace_name, pass_name)
+    # TODO: Also paste out the backtrace when the trace was compiled
+    # (and maybe also when it was run?)
+    with open(filename + ".ir", "w") as f:
+        f.write("Input key: {}\n\n{}".format(input_key, str(trace)))
+    graph_vis.write(trace.graph(), filename + ".html")
 
 
-def trace(**kwargs):
-    return lambda traceable: traced(traceable, **kwargs)
+@contextlib.contextmanager
+def _time(trace_name, name, time=True):
+    if (not _JIT_TIME and not time) or not torch.cuda.is_available():
+        yield
+        return
+    stream = torch.cuda.current_stream()
+    start = torch.cuda.Event(enable_timing=True)
+    end = torch.cuda.Event(enable_timing=True)
+    stream.record_event(start)
+    try:
+        yield
+    finally:
+        stream.record_event(end)
+        end.synchronize()
+        print("{} {} time: {} ms".format(trace_name, name, start.elapsed_time(end)))
```

Comment:
You're aware that this only records CUDA time, so it will show 0 if your model is CPU only, right?

Reply:
No, I was not. But this was the pre-existing timing code that I restored, so we should fix it in another PR.